thrift/client.go

```diff
@@ -155,6 +157,37 @@ func (c *client) Call(ctx Context, thriftService, methodName string, req, resp t
 		return false, err
 	}
 
-	ctx.SetResponseHeaders(respHeaders)
+	ctx.SetResponseHeaders(removeRPCServiceHeader(respHeaders))
 	return isOK, nil
 }
+
+// removeRPCServiceHeader removes the $rpc$-service from headers map when present
+// tchannel inbound does not add a $rpc$-header. tchannel outbound sets the response headers
+// that are received from the outbound call on to the context and they get sent back to the
+// caller of this inbound. This becomes problematic when a tchannel inbound server is an
+// intermediate part of a call graph where the initial callee verifies $rpc$-service header
+// Eg. A -> B -> C
+//           |-> D
+// In this call graph, let us say B is a tchannel inbound server and makes two tchannel outbound
+// calls to C and D. If C and D both reply with response headers, the headers get set in B's
+// context, one overwriting the other, and they get sent back to A. If either one of C or D or both
+// send a $rpc$-header, A will start receiving one of C's or D's $rpc$-service response header and
+// can perform validation on the header value and can inadvertently believe that some other service
+// other B handled the request that it sent to B and can mark the rpc as failure. This is the behavior
+// that is observed with yarpc/yarpc-go framework when one of the intermediaries in the call graph
+// is using a tchannel inbound server using a tchannel outbound.
+//
+// We have observed this issue while migrating to yarpc from tchannel inbound. For example, let us
+// assume that both C & D were tchannel inbounds earlier and one of them wants to migrate to yarpc
+// and at which point the $rpc$-service header flows upstream and assuming A is already using yarpc outbound
+// the call will fail. Although it is a backward incompatible change, by removing the $rpc$-service header
+// prevents A from seeing the $rpc$-header and marking the rpc as a failure. The alternative is to make
+// B use Child() context while making downstream calls and the net effect is the same and while migrating to yarpc,
+// it is unreasonable to expect the callee service owners to perform changes into the upstream callers
+// to create Child() context especially when the call graph is complex and multiple levels deep. This is
+// a reasonbable backward incompatible change to make as we do not expect any caller to actually depend on
+// or find the $rpc$-service header from a descendant service to be actually useful.
```

Comment:
Could you explain what you mean by "actually useful"?

It sounds like $rpc$-service was added there for a reason, and it's working as expected?

Reply:
$rpc$-service header is not added by tchannel inbound. It is added by YARPC inbound.

Sample topology: 
```
A (YARPC inbound/YARPC outbound) -> B (tchannel inbound/tchannel outbound) -> C (tchannel inbound)
                                                                          |-> D (tchannel inbound)
```

A makes a yarpc outbound call to B which makes two calls to C & D in parallel. None of C, D & B add $rpc$-service response header and so A will not receive one and so there is no validation from A. 

Now let's say C tries to upgrade to YARPC inbound. It starts emitting $rpc$-service response header.  Because tchannel outbound client in B sets the response headers that it receives from the outbound into its own context. The $rpc$-service header will propagate to A through B. A will see the $rpc$-service response header and finds that the response header is from C and not B as it expects and fails its rpc inbound with an error 

` service name sent from the request does not match the service name received in the response`

another wrinkle: As C & D are made in parallel by B, the response headers from C & D overwrite each other in B's context.

This is what I mean by the "$rpc$-service header from a descendant service" is not useful because it actually causes harm as A would inadvertently think that C is handling its request. 

The recommended solution from tchannel is to use Child() context while making outbound calls. ie. B should use Child() for all the outbound calls. The effect of that is that the downstream response headers are cutoff from being relayed upstream. What this means though is that while upgrading to YARPC, the service needs to find all its callers and modify all the codepaths to use a Child().

What this PR does is to make that process easier for service owner. As we see propagating $rpc$-service upstream from a tchannel outbound call causes unnecessary harm and using Child() anyway removes those headers from propagating upstream, we drop the $rpc$-service header from propagating upstream even when Child() is not used. It is also quite unreasonable to put that burden on the service owner to find out all its callers and modify the caller code. 

This change is only for tchannel outbound calls (to help us prevent outages). The validation in a yarpc only topology is still preserved.  ie A -> B where both A and B are YARPC outbound and inbound respectively, A will get a $rpc$-service response header and will validate.

Reply:
Ah, I misunderstood. My (apparently wrong) impression was that yarpc would be re-using tchannel-go, and so this PR would effectively disable `$rpc$-service` for yarpc as well. 

If this is not the case, I was wrong - sorry for the extra noise.

Reply:
I verified that yarpc does not use thrift.Client to make the outbound call. It uses tchannel [directly](https://github.com/yarpc/yarpc-go/blob/dev/transport/tchannel/channel_outbound.go#L174) and reimplemented most of the thrift.Client logic. 

If we want to be absolutely safe, we can actually make a copy of the map instead of directly modifying it. So what goes into the thrift.Context is a copy of the headers without the $rpc$-service.